torch/csrc/jit/codegen/cuda/python_frontend/README.md

```diff
@@ -0,0 +1,119 @@
+
+# Usage
+
+## Example
+
+```python
+import torch
+from torch._C._nvfuser import FusionManager, FusionDefinition, DataType
+
+fm = FusionManager.get()
+with FusionDefinition(fm) as fd :
+    t0 = fd.define_tensor(symbolic_sizes=[-1, 1, -1],
+                          contiguous=[True, True, True],
+                          dtype=DataType.Float)
+    t1 = fd.define_tensor(3)
+    c0 = fd.define_constant(3.0)
+
+    t2 = fd.ops.add(t0, t1)
+    t3 = fd.ops.mul(t2, c0)
+    t4 = fd.ops.sum(t3, [-1], False, DataType.Float)
+
+    fd.add_output(t4)
+
+input1 = torch.ones(2, 1, 8, device='cuda')
+input2 = torch.ones(2, 4, 8, device='cuda')
+
+nvf_out1 = fm.execute([input1, input2])[0]
+```
+
+## Components
+
+### `FusionManager` - Executes and Caches Defined Fusions
+#### `FusionManager` Methods
+* `get()`: Provides a global instance.  You cannot directly construct an instance of the `FusionManager`.
+* `execute([inputs])`:  Allows you to execute the currently defined fusion with a list of given inputs.
+* `print_ir()`: Prints the low level IR for the currently defined fusion.
+
+### `FusionDefiniton` Context Manager - Interface for Defining Fusions
+
+#### Defining Input Tensors
+_All intermediate tensors are created by operations.  Constant tensors do not exist._
+
+There are 3 ways to define tensors that will be enumerated below.
+
+##### 1.) Defining tensors by the number of input dimensions only
+This interface tells nvFuser that the tensor has a given number of symbolic dimensions that are not necessarily contiguous in memory.  The user also has the ability to specify a data type.  The default type is `Float`.
+```python
+t0 = fd.define_tensor(3)
+t1 = fd.define_tensor(3, DataType.Half)
+```
+
+##### 2.) Defining tensors by a list of concrete sizes and a list of strides
+The `sizes` parameter defines the number of dimensions and the size of each dimension.  The `strides` parameter has to have the same number of dimensions as the `sizes` parameter.
+nvFuser translates the concrete sizes and strides into symbolic sizes and contiguity information that can be directly defined via the next way to define tensors.  This allows the user to directly take a Pytorch defined tensor and query its sizes and strides in order to apply them in the definition.
+```python
+t0 = fd.define_tensor(sizes=[2, 4, 6], strides=[24, 6, 1], dtype=DataType.Half)
+```
+
+##### 3.) Defining tensors by a list of symbolic sizes and a list of contiguity information
+The list of symbolic sizes defines the number of dimensions and `-1` is given for each dimension unless it is a broadcast dimension that is defined with a `1`.  The contiguity information is viewed from right to left.  A `True` definition indicates the current dimension is contiguous with the dimension to its right.
+
+```python
+t0 = fd.define_tensor(symbolic_sizes=[-1, 1, -1], contiguous=[True, True, True], dtype=DataType.Float)
```

Comment:
What is the difference between `symbolic_sizes` and `sizes`? Is it that, if I want to specify `strides`, I have to use `sizes`?

Reply:
`symbolic_sizes` defines a dimension as symbolic (`-1`) or as a broadcast dimension (`1`).  The normal `sizes` version gives concrete sizes.

Reply:
What if I do `symbolic_sizes=[-1, 1, 10]`, is this an error?

Reply:
It is an error.  The python binding of `define_tensor()` that supports symbolic_sizes throws an assert.

https://github.com/kevinstephano/pytorch/blob/b9b4cb133519267577274082a846c78542b0b606/torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp#L156-L162